keps/sig-api-machinery/20200309-consistent-resource-versions-semantics/README.md

```diff
@@ -0,0 +1,236 @@
+	//   the watch started at;
+	// - if non-zero, start a watch at an exact resource version. The watch events are for all
+	//   changes after the provided resource version. Watch is not started with synthetic “Added”
+	//   events for the provided resource version. The client is assumed to already have the
+	//   initial state at the starting resource version since the client provided the resource version.
+	//
+	// ResourceVersion for list:
+	// - if unset, then the result is returned at the most recent resource version. The returned
+	//   data must be consistent (i.e. served from etcd via a quorum read);
+	// - if set and ResourceVersionMatch is set, ResourceVersion is applied according to the ResourceVersionMatch rule;
+	// - if set and ResourceVersionMatch is unset or the server ignores ResourceVersionMatch, the legacy behavior applies:
+	//   - if 0, the result may be at any resource version. The newest available resource version
+	//     is preferred, but strong consistency is not required; data at any resource version may
+	//     be served. It is possible for the request to return data at a much older resource version
+	//     that the client has previously observed, particularly in high availability configurations,
+	//     due to partitions or stale caches. Clients that cannot tolerate this should not use this
+	//     semantic;
+	//   - if non-zero and Limit is unset, the ResourceVersionMatchNotOlderThan rule applies implicitly;
+	//   - if non-zero and Limit is set, the ResourceVersionMatchExact rule applies implicitly.
+	//
+	// Defaults to unset
+	// +optional
+	ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,4,opt,name=resourceVersion"`
+
+	// ResourceVersionMatch determines how ResourceVersion is applied to list calls.
+	// It is highly recommmend that ResourceVersionMatch be set for list calls where ResourceVersion is
+	// set. If ResourceVersion is unset, ResourceVersionMatch is ignored.
+	// For backward compatibility, clients must tolerate the server ignoring ResourceVersionMatch:
+	// - When using ResourceVersionMatchNotOlderThan and Limit is set, clients must handle HTTP 410 “Gone” responses.
+	//   For example, the client might retry with a newer ResourceVersion or fall back to ResourceVersion="".
+	// - When using ResourceVersionMatchExact and Limit is unset, clients must verify that the ResourceVersion in the
+	//   ListMeta of the response matches the requested ResourceVersion, and handle the case where it does not. For
+	//   example, the client might fall back to the a request with Limit set.
+	// Unless you have strong consistency requirements, using ResourceVersionMatchNotOlderThan and a known
+	// ResourceVersion is preferable since it can achieve better performance and scalability of your cluster
+	// than leaving ResourceVersion and ResourceVersionMatch unset, which requires quorum read to be served.
```

Comment:
Thanks. I obviously would like to have higher in this comment, but I'm probably biased.
@lavalamp - WDYT?

Reply:
Happy to move this around either now or during PR code review. Also note that I rewrote the beginning of the `ResourceVersion` godoc to highlight the performance/scale implications.

Reply:
Given the other stuff this comment has to say, I'm not sure the performance bit can go higher and make sense :)

Reply:
Yeah - that's why I said "I'm probably biased".
Resolving - we can revisit in the future if needed.